### ISSUE TITLE
Background is not drawn if an inner stack has height

### ISSUE THREAD

**Issue by cyberarm** [shoes/shoes4#1021]
Tested on 8d392f47e41d4e3b147dbe6346291933a0927c1d (Shoes4 pre3)

Example of the failure:
``` ruby
Shoes.app do
  stack do
    stack height: '100%' do
      background gray
      para "content"
    end
  end
end
```

Example where it works as expected:
``` ruby
Shoes.app do
  stack do
    stack do
      background gray
      para "content"
    end
  end
end
```

Windows 7, JRuby 1.7.16.1.

Gems:
```
  * after_do (0.3.1)
  * bundler (1.7.6)
  * furoshiki (0.3.0)
  * jruby-jars (1.7.18)
  * jruby-rack (1.1.17)
  * nokogiri (1.6.4.1)
  * plist (3.1.0)
  * rake (10.4.2)
  * rubyzip (1.1.6)
  * shoes (4.0.0.pre3)
  * shoes-core (4.0.0.pre3)
  * shoes-highlighter (1.0.0)
  * shoes-manual (4.0.0)
  * shoes-package (4.0.0.pre3)
  * shoes-swt (4.0.0.pre3)
  * swt (4.4)
  * warbler (1.4.4)
```

**Comment by jasonrclark** [comment]
Thanks for checking out `pre3` and thanks for the report. Will definitely look into this.

**Comment by KCErb** [comment]
I see the same bug on OSX, also I couldn't find a commit where this works. 

I think I shall do some digging!

**Comment by KCErb** [comment]
Hmm, I'm getting the background drawn if I give the outer stack content, but when I resize the app it hangs :(

```ruby
Shoes.app do
  stack do
    para "more content"
    stack height: '100%' do
      background gray
      para "content"
    end
  end
end
```

I also can't find the commit where this was introduced. I went back as far as where string handling for dimensions was added. Hmm

**Comment by PragTob** [comment]
confirmed & working on this (initial problem as well as the hanging problem :( ) initial working theory - the parent stack doesn't have height due to no content except for the stack, that derivates its content from the parent stack... and 100% of 0 is 0 :)

**Comment by PragTob** [comment]
Okay so this is like I expected it to be... height: '100%' means 100% of the parent height and since the parent stack is a normal stack, which means its height its determined by its contents and the only content is a stack with the height 0... it is 0.

Interesting to see what the correct/good behavior here might be. Shoes 3.2 just shows nothing - which makes sense as both stacks have the height 0. We still paint the `content` as we don't check if what we paint is outside of the contaienr contents... sigh. 

Any ideas about what the behavior should be around here?

**Comment by PragTob** [comment]
The hang @KCErb discovered is one of my "favorite" places of the code base - the Shoes layout! Ot at least most likely so... I know @jasonrclark fixed a hang somewhere around there some time ago as well.

**Comment by jasonrclark** [comment]
So, you want to do #994 next or shall I? :grin: 

FWIW, my last hang started out looking like it was in `ShoesLayout`, but ended up actually being in the SWT `layout` calls when we were doing `contents_alignment`.

**Comment by PragTob** [comment]
If I get down to it I'll probably write a spec... maybe not today but I'll see right now I'm still at "tracking this stuff down" 

I'm also asking myself again why we need a layout at all or why it needs to be special as we do the positioning and stuff with slots... hmm.

FYI the hang comes to life as it appears that something starts to want to grow continously:

Values displayed are first current container size, and then desired next container size as computed by `set_gui_size`:

```
606x547
606x547
----------------------
606x547
606x560
----------------------
606x560
606x580
----------------------
606x580
606x600
----------------------
606x600
606x620
----------------------
606x620
606x640
----------------------
606x640
606x660
----------------------
606x660
606x680
----------------------
606x680
606x700
----------------------
606x700
606x720
----------------------
606x720
606x740
----------------------
606x740
606x760
----------------------
606x760
606x780
----------------------
(it never stops)
```

**Comment by jasonrclark** [comment]
Interesting. Reminds me of d2933c23d106a46fcac814ce1e84cfb0246637db where `background` and `border` would cause that, but those weren't supposed to be included in the height calculation and everything in this test app belongs.

**Comment by PragTob** [comment]
So here's what's happening, the inner and outer stack play ping pong with each other as they both grow each other...

simplified the code is like this:

```ruby
Shoes.app do
  stack do
    para 'I got content'
    stack height: '100%' do
      background green
    end
  end
end
```

We start out with the outer stack at height 20px (due to the para). Then the inner stack has a height of 20px too. The next time we call contents alignment the outer stack has a height of 40 px (para + inner stack) and hence the inner stack gets a height of 40px. Next time around the outer stack has a height of 60px (20px para + inner stack) and so has the inner stack... and so on and so forth. It never ends and keeps resizing itself.

I'm not sure what to do about this. The most sensible solution I can think of is that parent slots should not take any dimension into account when calculating their own dimensions that are set relative to themselves. But that gets hard... what if there are multiple children of a slot, some relative, some not... argh.

Ideas?

**Comment by wasnotrice** [comment]
The inner stack's height should be lazy, no? In other words, because its height is *relative to the parent*, iit should calculate its height based on the parent, and not the other way around. So in this case:

```
para: I'm 20px
inner stack: I don't know how tall I am yet
outer stack: I'm 20px
inner stack: I'm 20px
```

**Comment by jasonrclark** [comment]
The more I've thought about this today, the more confused I got about what the desirable behavior even was. Good times :grin:

I started writing out an exhaustive matrix relating the outer/inner sizing, but partway through I had an idea how to simplify things.

The crux of the issue is a circular dependency between outer and inner slot sizes. So I say we simply outlaw that--disallow outer slots depending on the content size *when those contents include relatively sized slots*.

So how do we react when an app break the "rule" (like @PragTob's above?) Well, just treat the outer slot as "100%" of *its* parent instead sizing it to the child contents. This would need some basic looking ahead in the slot positioning, but a slot could definitively answer its own size before inner relatives need it.

In @PragTob's test app above you'd see the text, followed by the green slot consuming the rest of the window (or as tall as the window, so scrolling off a bit? Having a slot at 100% the parent size as a sibling to another element is odd). That seems like a sane result to me. And if you're not happy with this promotion to 100% sizing? Well, just give your slot the size you actually want! :stuck_out_tongue: 

To summarize the proposal: slots with relatively sized children don't size to their contents, they size to their parents.

A final note... in Shoes 3, if you run the test app everything looks ok at first--the slot gets a size of ~20px like the text above it. But if you resize the window, the scrollbar shoots off really long. Basically, I think it has the same problem, it just doesn't hang from it.

**Comment by PragTob** [comment]
@wasnotrice I like the idea, I thought of something similar when I said that the size of relative stacks gets ignored - it just has the slight problem that they already have a size set from the first run through but we can check if their height is relative I guess. That should/could work :) 

@jasonrclark slots with relative sized children don't size to their contents but to their parents? Interesting idea... I just don't get how it is supposed to work... so say:

```ruby
slot do
  slot do
    slot height: '100%'
    # other contents
  end
  slot do
    slot height: '100%'
    # other contents
  end
end
```

the two middle slots can hardly size to heir parent, as then both would be the size of the parent... maybe I'm just missunderstanding it, though.

Summarizing: I think I don't understand what @jasonrclark wrote - please explain it :) And I think that what @wasnotrice wrote could work :) But I'm not sure. All pending testing! :)

**Comment by jasonrclark** [comment]
Hmm, good point about the intermediates. Guess I was more tired than I thought :( Re-reading through my idea, it seems like this probably won't work.

That being said, what does it mean for a slot to be 100% of it's parent height, but then be followed by other content? It can't fill the slot entirely because there are other contents... but what other size should it get?

I still had some question with the laziness about what happens the redraw, but now I think I get it. Next time it'd do the same lazy calculation, so outer would still stick with 20px. If I'm getting that, seems like it'd work.

Sorry for the long diversion there!

**Comment by PragTob** [comment]
However, I think I don't get how it is supposed to work :)

Let's look at this:

```ruby
stack do
  para 'para 1'
  stack height: 100%
  para 'para 2'
end
```

Question during positioning - where does `para 2` get positioned? We don't know yet how big the parent slot is - so if we just skip it `para 2` gets positioned directly after `para 1`... which is wrong but otherwise we don't know what the size of the stack should be.

The only thing I can imagine is that we at first ignore the stack size... and once we positioned the last element we go back and position all the relatively positioned elements (as we now know their height) and then reposition the other elements accordingly... sounds a bit awkward :(

**Comment by KCErb** [comment]
### WARNING: Long winded discussion of my *ideas*, might be more fluff than content. 

<hr>

One of the very first things that confused me when first learning shoes was: 
* What is a fixed height stack?
* What is a fixed width flow?

I haven't dug into Layout (I will someday!) so my opinion is more based on my reading of the manual / NKS. Also my very first ruby program was a very big, and very slot intensive app all written for Shoes 3 so I have a funny squishy feeling for stacks and flows from that. 

I learned in that process that if I wanted to have a box of fixed width *and* height then the outter most container should be a flow (I'm confident that that is why the app window itself is a flow). It always seemed contradictory to me to fix the height of a stack. The whole point of a stack is to be a container that runs vertically to match its content. Likewise a fixed-width flow seems unnatural, but a fixed-width flow is not nearly as bad as a fixed height stack.

Why not so bad? Because of what `_why` said in Nobody Knows Shoes: 

> I'm sure the question you're wondering now is: shouldn't there just be **vertical** stacks and **horizontal** stacks??<br>
But the thing is: Shoes doens't believe in a horizontal SCROLLBAR!!!<br>
So flows hit the end and move down...

So the horizontal stack is replaced by the flow which intrinsically has a fixed width: the window size.

The point is, if I fill a flow with content it will wrap around and down, it won't extend off the page. I can cheat and tell the flow to have a fixed width greater than the window size, and in which case *it will* flow off the page horizontally:

```ruby
Shoes.app width: 100 do
  flow width: 200 do
    para "para 1 is longer now"
  end
end
```

But that breaks what flows are all about! The normal use of a flow is to have the width free to match the window. 

Stacks on the other hand *can* extend off the page, vertically. So when I think of a stack I think of something that starts out with infinite height and then gets trimmed down based on the contents so that the computer doesn't explode. When I think of a flow I think of something as wide as the page that slowly grows its height to match its contents.

I don't know how well that paradigm matches the current implementation of Layout (yet) but it does answer some questions (for me at least).

In that paradigm then the answer to the question about where para 2 should go in @PragTob's example is simple: I stacked 3 things 
1. A thing with a fixed height: para
2. A thing with infinite height: stack height: '100%'
3. a thing with fixed height para.

So the result is an infinite column. In my opinion an app like that should (gracefully) break. It should be trying to add something to infinity and failing. You get similar behavior running the following app in Shoes 3 and 4:

```ruby
Shoes.app do
  stack height: '100%'
  stack height: '100%'
end
```

I think the infini-scroll behavior here is expected, but not observed in the case of @PragTob's app. 

Now to be clear it should **not** break if I just have 1 `stack height: '100%'`  because there's no problem being 100% the height of the window since the window also has a theoretical infinite height (remember the possibility to scroll forever).

<hr>
**Sidenote**: It's always bothered me that stacks have a default width of 100%. A 100% high stack should be a skinny column, potentially with width 0px if there's no content. We should save the behvaior where I automatically get 100% width for flows.
<hr>

What about the rest of the example apps in this thread? They all are asking for too much vertical space. 100% means all of the space. If you want A to take up 100% of the available vertical space then add something to the top or bottom of A you should get :boom:. 

The last question I need to address though is: what about relative sizes? The problems these apps are running into is when you have a stack within a stack. Let's take the opening question:

```ruby
Shoes.app do
  stack do
    stack height: '100%' do
      background gray
      para "content"
    end
  end
end
```

I think of it like this: So I want to stack some things, the first 1 is 100% the height of it's parent which is a plain stack. Plain stacks start out at infinity (window height + scrolling potential, think twitter) and get trimmed later. So in that infinitely high container, I now place a background that fills it up and next a para. After all of those are queued up, the stacks get trimmed down to match their contents. In this case means:

```
app (which is a flow) -> default app height
  stack -> height of contents
    stack -> default app height
```

So for someone who still feels more like a shoes user than a shoes developer that's where I'm at with expected / ideal behavior in these edge cases. But I recognize that my reasoning may be faulty, or just bad and I'm happy to support any solutions you guys cook up. :D

**Comment by jasonrclark** [comment]
Interesting ideas and examples. Appreciate your ideas @KCErb, because I'm actually pretty keen on the user perspective for how those values play together, and I just don't have my own head around what I would expect beyond the most basic cases.

But that said, it's been a busy, brain-full week, so I probably won't get a lot of thinking done until the weekend. Maybe then I'll be rested enough to grapple with the slots! :grin:

**Comment by PragTob** [comment]
Looks like besides the packaging issues this is the only thing remaining for pre4... I'll see that I give it a shot on the weekend. Although I'm still not 100% clear on how to handle this...

**Comment by jasonrclark** [comment]
Yeah, I've been avoiding this one for a while now. :stuck_out_tongue_closed_eyes:

After reading and re-reading I'm **still** not sure of the behavior in all the examples presented here.

Doing a separate, later pass to fill in relative values seems plausible technically, but the results seem likely to surprise to me. For instance, @wasnotrice's take on the original example with lazy values suggests to me we'd see `para "content"` with either nothing or a 20px background below it. Neither of those seem like what I'm asking for by setting that `100%`.

A half-baked thought, but what if relative sizes just keep looking up the slot parent chain until they find an actual value? In this example, you'd end up with a large background, 100% of the window size, but that seems more aligned with what I'd expect. That also plays nicely with siblings asking different relative amounts, since they'd both resolve to the same parent to calculate the value against.

Will also put out there, I'm fine with postponing this to `rc1` in favor of getting `pre4` out. I put it in `pre4` because at the time the topic was fresh in my mind. Now it's less fresh, but no clearer for the break.

**Comment by jasonrclark** [comment]
Looks like the packaging issues are getting themselves cleared up. How would you feel about a bump to [pre5](https://github.com/shoes/shoes4/issues?q=is%3Aopen+is%3Aissue+milestone%3A4.0.0.pre5) on this @PragTob?

**Comment by KCErb** [comment]
Two cents: this is a sticky topic and pre4 has come along nicely enough to do fine without it. I think some reasonable suggestions for how to handle this are floating around. As long as we end up somewhere that can produce predictable behavior (such as "look up the chain until you find a value") then we'll have brought this to a good place and that's better than where we are :stuck_out_tongue:

**Comment by PragTob** [comment]
It's cool to bump this, if someone gets to it before pre4 is released well then it's magic and we're all good.... but true , it is a weird setup...

Looking back at the sample:

```
Shoes.app do
  stack do
    para 'I got content'
    stack height: '100%' do
      background green
    end
  end
end
```

What I'd expect would be that the outer stack is 40px (given that the text block takes up 20px) and then inner stack is also 20 px... and it stays like that.

Now what did the user expect? I have no idea. Maybe a green background behind the stack and the text - but then doing a `background` call right in that stack would be the right thing to do. 

Thinking of percent values and slots and when I use them - usually for layouts.

```
Shoes.app do
  stack height: 1.0 do
    flow height: 1.0 do
      stack width: 0.2, height: 1.0 do
        background yellow
        title 'L'
      end
      stack width: 0.6, height: 1.0 do
        flow height: 0.2 do
          background green
          title 'H'
        end
        stack height: 0.8 do
          background blue
          title 'Main'
        end
      end
      stack width: 0.2, height: 1.0 do
        background purple
        title 'R'
      end
    end
  end
end
```
![screenshot from 2015-05-05 11 49 35](https://cloud.githubusercontent.com/assets/606517/7470402/15ddc708-f31d-11e4-8810-13049fb7f7f2.png)

This makes me believe that they should take up the space they claim to take up.. trouble enters paradise when there are other elements in the stack and you say you want 100% of the space, which in the end is more space than there is... aw, headeche :(

**Comment by jasonrclark** [comment]
Thanks for the bump @PragTob! Will continue to mull this over whenever I want a good headache :grin:

**Comment by jasonrclark** [comment]
Any objections to putting this back for rc1? I don't think I'm going to be able to get to it before I'd like to cut a pre5 release. @PragTob @KCErb ?

**Comment by PragTob** [comment]
It's fine.. it seems like this may take much longer than initially expected.

**Comment by KCErb** [comment]
:+1:
